batching_kafka_consumer/__init__.py

```diff
@@ -235,39 +243,50 @@ def _shutdown(self):
 
     def _reset_batch(self):
         logger.debug("Resetting in-memory batch")
-        self.uncommitted_messages = False
-        self.batch = []
-        self.timer = None
+        self.__batch_results = []
+        self.__batch_deadline = None
+        self.__batch_messages_processed_count = 0
+        self.__batch_messages_processed_duration_ms = 0.0
 
     def _flush(self, force=False):
         """Decides whether the `BatchingKafkaConsumer` should flush because of either
         batch size or time. If so, delegate to the worker, clear the current batch,
         and commit offsets to Kafka."""
-        if self.uncommitted_messages:
-            batch_by_size = len(self.batch) >= self.max_batch_size
-            batch_by_time = self.timer and time.time() > self.timer
-            if (force or batch_by_size or batch_by_time):
-                logger.info(
-                    "Flushing %s items: forced:%s size:%s time:%s",
-                    len(self.batch), force, batch_by_size, batch_by_time
+        if not self.__batch_messages_processed_count > 0:
+            return  # No messages were processed, so there's nothing to do.
+
+        batch_by_size = len(self.__batch_results) >= self.max_batch_size
+        batch_by_time = self.__batch_deadline and time.time() > self.__batch_deadline
+        if (force or batch_by_size or batch_by_time):
+            logger.info(
+                "Flushing %s items: forced:%s size:%s time:%s",
+                len(self.__batch_results), force, batch_by_size, batch_by_time
+            )
+
+            if self.metrics:
+                self.metrics.timing(
+                    'process_message.normalized',
+                    self.__batch_messages_processed_duration_ms / self.__batch_messages_processed_count,
                 )
 
-                if self.batch:
-                    logger.debug("Flushing batch via worker")
-                    t = time.time()
-                    self.worker.flush_batch(self.batch)
-                    duration = int((time.time() - t) * 1000)
-                    logger.info("Worker flush took %sms", duration)
-                    if self.metrics:
-                        self.metrics.timing('batch.flush', duration)
-
-                logger.debug("Committing Kafka offsets")
-                t = time.time()
-                self._commit()
-                duration = int((time.time() - t) * 1000)
-                logger.debug("Kafka offset commit took %sms", duration)
-
-                self._reset_batch()
+            batch_results_length = len(self.__batch_results)
```

Comment:
Also: I pulled this out because `worker.flush_batch` could hypothetically mutate this list, changing the number of items.